controllers/volsync/vshandler.go

```diff
@@ -1571,3 +1662,376 @@ func ValidateObjectExists(ctx context.Context, c client.Client, obj client.Objec
+	lrs := &volsyncv1alpha1.ReplicationSource{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      getLocalReplicationName(rd.GetName()),
+			Namespace: v.owner.GetNamespace(),
+		},
+	}
+
+	err = v.client.Get(v.ctx, types.NamespacedName{
+		Name:      lrs.Name,
+		Namespace: lrs.Namespace,
+	}, lrs)
+	if err != nil {
+		if !kerrors.IsNotFound(err) {
+			v.log.Error(err, "Unable to get Local ReplicationSource", "LocalRS", lrs)
+
+			return nil, fmt.Errorf("error getting Local ReplicationSource (%w)", err)
+		}
+	}
+
+	err = v.cleanupPreviousTransferResources(lrs, vsImageRef.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	snap, err := v.validateSnapshotAndAddDoNotDeleteLabel(*vsImageRef)
+	if err != nil {
+		return nil, err
+	}
+
+	var restoreSize *resource.Quantity
+
+	if snap.Status != nil {
+		restoreSize = snap.Status.RestoreSize
+	}
+
+	// In all other cases, we have to create a RO PVC.
+	return v.createReadOnlyPVCFromSnapshot(rd, *latestImage, restoreSize)
+}
+
+//nolint:gocognit,nestif
+func (v *VSHandler) cleanupPreviousTransferResources(lrs *volsyncv1alpha1.ReplicationSource,
+	snapImageName string,
+) error {
+	if lrs.Spec.Trigger != nil && lrs.Spec.Trigger.Manual != snapImageName {
+		// Only clean up and create new PVC if the previous transfer has completed. We don't want to abort it.
+		if lrs.Spec.Trigger.Manual != "" {
+			if lrs.Status != nil && lrs.Status.LastManualSync == lrs.Spec.Trigger.Manual {
+				err := v.deleteSnapshot(v.ctx, v.client, lrs.Spec.Trigger.Manual, v.owner.GetNamespace(), v.log)
+				if err != nil {
+					return err
+				}
+
+				err = util.DeletePVC(v.ctx, v.client, lrs.Spec.SourcePVC, v.owner.GetNamespace(), v.log)
+				if err != nil {
+					return err
+				}
+			} else { // don't process any further. We have to wait for the previous transfer to complete
+				return fmt.Errorf("wait for previous localRS tranfter to complete - RS/PVC %s/%s",
+					lrs.GetName(), lrs.Spec.SourcePVC)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (v *VSHandler) createReadOnlyPVCFromSnapshot(rd *volsyncv1alpha1.ReplicationDestination,
+	snapshotRef corev1.TypedLocalObjectReference, snapRestoreSize *resource.Quantity,
+) (*corev1.PersistentVolumeClaim, error) {
+	l := v.log.WithValues("pvcName", rd.GetName(), "snapshotRef", snapshotRef,
+		"snapRestoreSize", snapRestoreSize)
+
+	pvc := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      snapshotRef.Name,
+			Namespace: v.owner.GetNamespace(),
+		},
+	}
+
+	pvcRequestedCapacity := rd.Spec.RsyncTLS.Capacity
+	if snapRestoreSize != nil {
+		if pvcRequestedCapacity == nil || snapRestoreSize.Cmp(*pvcRequestedCapacity) > 0 {
+			pvcRequestedCapacity = snapRestoreSize
+		}
+	}
+
+	op, err := ctrlutil.CreateOrUpdate(v.ctx, v.client, pvc, func() error {
+		if pvc.Status.Phase == corev1.ClaimBound {
+			// PVC already bound at this point
+			l.V(1).Info("PVC already bound")
+
+			return nil
+		}
+
+		accessModes := []corev1.PersistentVolumeAccessMode{corev1.ReadOnlyMany}
```

Comment:
ModifyRSSpecForCephFS() should be setting the access modes we need right? And that's already run prior to calling this function?  Only cephfs needs the "read-only" PVC from snapshot here - we're making a PVC from snapshot and the rsSpec should already have been modified by modifyRSSpecForCephFS():  https://github.com/RamenDR/ramen/blob/cac8089eb79fec6875bb52c2c8ca532780b1c615/controllers/volsync/vshandler.go#L1228

To be consistent maybe we should take values from the rsSpec and let it decide whether it's read-only - not that anyone should be using LocalDirect for storageclasses that aren't cephfs